pkg/gateway/middleware/inject.go

```diff
@@ -0,0 +1,30 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/skygeario/skygear-server/pkg/core/inject"
+	"github.com/skygeario/skygear-server/pkg/gateway"
+)
+
+// InjectableMiddleware is a pointer to a struct
+// with `dependency:` tags.
+type InjectableMiddleware interface {
+	Handle(next http.Handler) http.Handler
+}
+
+// Injecter injects dependencies from Dependency into
+// Middleware at request time.
+type Injecter struct {
+	Middleware InjectableMiddleware
+	Dependency gateway.DependencyMap
+}
+
+// Handle implements gorilla middleware signature.
+func (m Injecter) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		inject.DefaultRequestInject(m.Middleware, m.Dependency, r)
```

Comment:
Since gateway is multi-tenant, it is not safe to inject dependency to the same middleware instance and use it for all the requests. If we want to reuse the core inject function, we will need to do it like core. We should use the middleware factory, the middleware instance will be created in the request runtime and should not be reused.

Reply:
Fixed.